### PR TITLE
rgw: DefaultRetention requires either Days or Years

### PR DESCRIPTION
**Pull request by liuchang0812**
Either Days or Years must be specified, but not both.

see https://docs.aws.amazon.com/AmazonS3/latest/API/Type_API_DefaultRetention.html

Signed-off-by: Chang Liu <liuchang0812@gmail.com>

@cbodley @yehudasa mind taking a look?

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

File: src/rgw/rgw_object_lock.h [hunk 1]
```diff
@@ -106,7 +106,11 @@ class RGWObjectLock
   }
 
   bool retention_period_valid() const {
-    return (get_years() > 0 || get_days() > 0);
+    // DefaultRetention requires either Days or Years.
+    // You can't specify both at the same time.
+    // see https://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketPUTObjectLockConfiguration.html
+    return (get_years() > 0 || get_days() > 0) && 
+	   !(get_years() > 0 && get_days() > 0);
```
Comment (tchaikov) [inline src/rgw/rgw_object_lock.h]: could be
```c++
return get_years() > 0 != get_days() > 0;
```

File: src/rgw/rgw_object_lock.h [hunk 2]
```diff
@@ -106,7 +106,10 @@ class RGWObjectLock
   }
 
   bool retention_period_valid() const {
-    return (get_years() > 0 || get_days() > 0);
+    // DefaultRetention requires either Days or Years.
+    // You can't specify both at the same time.
+    // see https://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketPUTObjectLockConfiguration.html
+    return (get_years() > 0) != (get_days() > 0);
```
Comment (tchaikov) [inline src/rgw/rgw_object_lock.h]: nit, no need to put parentheses around the statement after `return`.

Comment (liuchang0812) [reply src/rgw/rgw_object_lock.h]: thanks, there is a warning if we remove those parentheses. 

``` 
/home/liuchang/WorkSpace/ceph/src/rgw/rgw_object_lock.h:112:24: warning: suggest parentheses around comparison in operand of ‘!=’ [-Wparentheses]
     return get_years() > 0 != get_days() > 0;
            ~~~~~~~~~~~~^~~
```

**Comment by ivancich** [comment]
rgw/(non-)verify: http://pulpito.ceph.com/ivancich-2019-09-11_01:23:37-rgw-wip-eric-testing-1-distro-basic-smithi/
rgw/verify: http://pulpito.ceph.com/ivancich-2019-09-13_17:22:14-rgw-wip-eric-testing-1-bisect-6-distro-basic-smithi/

**Comment by ivancich** [comment]
@liuchang0812 Can you create a tracker for this and cross-link them? It passed my QA and I'm ready to merge, but so we need the tracker so the backports can be tracked. Thanks!

**Comment by liuchang0812** [comment]
> @liuchang0812 Can you create a tracker for this and cross-link them? It passed my QA and I'm ready to merge, but so we need the tracker so the backports can be tracked. Thanks!

done, https://tracker.ceph.com/issues/41838, and I have updated pr's commit message and rebased to master.

**Comment by liuchang0812** [comment]
jenkins test make check

**Comment by cbodley** [comment]
jenkins test make check

**Pull request merged**